File: 133. Clone Graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, neighbors):
        self.val = val
        self.neighbors = neighbors
"""
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        '''
        DFS solution with stack
        '''
        if not node:
            return None
        newnode = Node(node.val, [])
        stack = [node]
        dic = {node: newnode}
        while stack:
            temp = stack.pop(-1)
            for n in temp.neighbors:
                if n not in dic:
                    dic[n] = Node(n.val, [])
                    dic[temp].neighbors.append(dic[n])
                    stack.append(n)
                else:
                    dic[temp].neighbors.append(dic[n])
        return newnode


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        '''
        BFS solution with queue
        '''
        if not node:
            return None
        newnode = Node(node.val, [])
        queue = [node]
        dic = {node: newnode}
        while queue:
            temp = queue.pop(0)
            for n in temp.neighbors:
                if n not in dic:
                    dic[n] = Node(n.val, [])
                    dic[temp].neighbors.append(dic[n])
                    queue.append(n)
                else:
                    dic[temp].neighbors.append(dic[n])
        return newnode

class Solution:
    '''
    DFS recursion
    '''
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        newnode = Node(node.val, [])
        self.dic = {node: newnode}
        self.dfs(node)
        return newnode
        
        
    def dfs(self, node):
        for n in node.neighbors:
            if n not in self.dic:
                self.dic[n] = Node(n.val, [])
                self.dic[node].neighbors.append(self.dic[n])
                self.dfs(n)
            else:
                self.dic[node].neighbors.append(self.dic[n])

File: 133. Clone Graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
/**
 * DFS iteration method
 * Time O(N * E)
 * Space O(N)
 */
class Solution {
    public Node cloneGraph(Node node) {
        HashMap<Node, Node> map = new HashMap<>();
        Deque<Node> stack = new ArrayDeque<>();
        HashSet<Node> seen = new HashSet<>();
        Node cnode = new Node(node.val, new ArrayList<>());
        map.put(node, cnode);
        stack.push(node);
        seen.add(node);
        Node head = node;
        while (!stack.isEmpty()) {
            node = stack.pop();
            for (Node topush : node.neighbors) {
                if (!seen.contains(topush)) {
                    seen.add(topush);
                    stack.push(topush);
                }
                if (map.containsKey(topush)) {
                    map.get(node).neighbors.add(map.get(topush));
                } else {
                    cnode = new Node(topush.val, new ArrayList<>());
                    map.put(topush, cnode);
                    map.get(node).neighbors.add(cnode);
                }  
            }
        }
        return map.get(head);
    }
}
/**
 * BFS iteration method
 * Same as DFS iteration
 */
class Solution {
    public Node cloneGraph(Node node) {
        HashMap<Node, Node> map = new HashMap<>();
        Deque<Node> queue = new ArrayDeque<>();
        HashSet<Node> seen = new HashSet<>();
        Node cnode = new Node(node.val, new ArrayList<>());
        map.put(node, cnode);
        queue.offerLast(node);
        seen.add(node);
        Node head = node;
        while (!queue.isEmpty()) {
            node = queue.pollFirst();
            for (Node topush : node.neighbors) {
                if (!seen.contains(topush)) {
                    seen.add(topush);
                    queue.offerLast(topush);
                }
                if (map.containsKey(topush)) {
                    map.get(node).neighbors.add(map.get(topush));
                } else {
                    cnode = new Node(topush.val, new ArrayList<>());
                    map.put(topush, cnode);
                    map.get(node).neighbors.add(cnode);
                }  
            }
        }
        return map.get(head);
    }
}
/**
 * DFS recursion
 * same as above
 */
class Solution {
    public Node cloneGraph(Node node) {
        HashMap<Node, Node> map = new HashMap<>();
        HashSet<Node> seen = new HashSet<>();
        dfs(node, map, seen);
        return map.get(node);
    }
    private void dfs(Node node, HashMap<Node, Node> map, HashSet<Node> seen) {
        seen.add(node);
        if (!map.containsKey(node)) {
            map.put(node, new Node(node.val, new ArrayList<>()));
        }
        for (Node next : node.neighbors) {
            if (!map.containsKey(next)) {
                map.put(next, new Node(next.val, new ArrayList<>()));
            }
            map.get(node).neighbors.add(map.get(next));
            if (!seen.contains(next)) {
                dfs(next, map, seen);
            }
        }
    }
}